File: javascript/metaparticle-package/metaparticle-runner.js
(function() {
    var shell = require('shelljs');
    var fs = require('fs');

    var ports = (options) => {
        var portArr = [];
        if (options && options.ports) {
            for (var ix = 0; ix < options.ports.length; ix++) {
                portArr.push({ "number": options.ports[ix] });
            }
        }
        return portArr;
    };

    module.exports.run = (img, name, options) => {
        var service = {
            "name": name,
            "guid": 1234567,
        };

        if (options && (options.replicas || options.shardSpec)) {
            service.services = [{
                "name": name,
                "replicas": options.replicas,
                "shardSpec": options.shardSpec,
                "containers": [
                    { "image": img }
                ],
                "ports": ports(options)
            }];
            service.serve = {
                "name": name,
            };
            if (options && options.public) {
                service.serve.public = true;
            }
        }
        if (options && options.jobSpec) {
            service.jobs = [{
                "name": name,
                "replicas": options.jobSpec.count,
                "containers": [
                    { "image": img }
                ]
            }];
        }
        if (!fs.existsSync(".metaparticle")) {
            fs.mkdirSync(".metaparticle");
        }
        var serviceString = JSON.stringify(service, null, '\t');
        fs.writeFileSync(".metaparticle/service.json", serviceString);

        // TODO: these should all be async so that the ctrl-c handling above works properly...
        shell.exec('mp-compiler -f .metaparticle/service.json');
        shell.exec('mp-compiler -f .metaparticle/service.json --deploy=false --attach=true');
    };

    module.exports.cancel = () => {
        // TODO: this should be async.
        var res = shell.exec('mp-compiler -f .metaparticle/service.json --delete');
        if (res.code != 0) {
            console.log('delete failed!');
            console.log(res.stdout);
            console.log(res.stderr);
        }
    };
})();

File: ruby/metaparticle/lib/metaparticle/metaparticle_runner.rb
require 'fileutils'

module Metaparticle
  class MetaparticleRunner
    def initialize(config)
      @config = config
    end

    def run
      options = {
        name: @config.name,
        guid: 1234567,
        services: [
          {
            name: @config.name,
            replicas: @config.replicas,
            shardSpec: @config.shardSpec,
            containers: [
              {
                image: @config.image
              }
            ],
            ports: @config.ports.map {|p| {number: p}}
          }
        ],
        serve: {
          name: @config.name
        }
      }
      if @config.public
        options[:serve][:public] = true
      end

      create_directory

      File.open(".metaparticle/service.json", "w") do |f|
        f.write(options.compact!.to_json)
      end

      `mp-compiler -f .metaparticle/service.json`
      `mp-compiler -f .metaparticle/service.json --deploy=false --attach=true`
    end

    private
    def create_directory
      FileUtils::mkdir_p('.metaparticle')
    end
  end
end
